// one/tmp/built_noprocess.js
/*global require:false, Buffer:false, process:false, module:false */
var exampleProject = (function(unused, undefined){
  var DEBUG         = false,
      pkgdefs       = {},
      pkgmap        = {},
      global        = {},
      lib,
      nativeRequire = typeof require != 'undefined' && require,
      nativeBuffer  = typeof Buffer != 'undefined' && Buffer,
      ties, locals;
  lib = (function(exports){
  exports.path = (function(exports){
    // Copyright Joyent, Inc. and other Node contributors.
// Minimized fork of NodeJS' path module, based on its an early version.
exports.join = function () {
  return exports.normalize(Array.prototype.join.call(arguments, "/"));
};
exports.normalizeArray = function (parts, keepBlanks) {
  var directories = [], prev;
  for (var i = 0, l = parts.length - 1; i <= l; i++) {
    var directory = parts[i];
    // if it's blank, but it's not the first thing, and not the last thing, skip it.
    if (directory === "" && i !== 0 && i !== l && !keepBlanks) continue;
    // if it's a dot, and there was some previous dir already, then skip it.
    if (directory === "." && prev !== undefined) continue;
    if (
      directory === ".." &&
      directories.length &&
      prev !== ".." &&
      prev !== "." &&
      prev !== undefined &&
      (prev !== "" || keepBlanks)
    ) {
      directories.pop();
      prev = directories.slice(-1)[0];
    } else {
      if (prev === ".") directories.pop();
      directories.push(directory);
      prev = directory;
    }
  }
  return directories;
};
exports.normalize = function (path, keepBlanks) {
  return exports.normalizeArray(path.split("/"), keepBlanks).join("/");
};
exports.dirname = function (path) {
  return path && path.substr(0, path.lastIndexOf("/")) || ".";
};
    return exports;
  }({}));
  return exports;
}({}));
  function findPkg(uri){
    return pkgmap[uri];
  }
  function findModule(workingModule, uri){
    var module,
        moduleId = lib.path.join(lib.path.dirname(workingModule.id), uri).replace(/\.js$/, ''),
        moduleIndexId = lib.path.join(moduleId, 'index'),
        pkg = workingModule.pkg;
    var i = pkg.modules.length,
        id;
    while(i-->0){
      id = pkg.modules[i].id;
      if(id==moduleId || id == moduleIndexId){
        module = pkg.modules[i];
        break;
      }
    }
    return module;
  }
  function genRequire(callingModule){
    return function require(uri){
      var module,
          pkg;
      if(/^\./.test(uri)){
        module = findModule(callingModule, uri);
      } else if ( ties && ties.hasOwnProperty( uri ) ) {
        return ties[ uri ];
      } else {
        pkg = findPkg(uri);
        if(!pkg && nativeRequire){
          try {
            pkg = nativeRequire(uri);
          } catch (nativeRequireError) {}
          if(pkg) return pkg;
        }
        if(!pkg){
          throw new Error('Cannot find module "'+uri+'" @[module: '+callingModule.id+' package: '+callingModule.pkg.name+']');
        }
        module = pkg.index;
      }
      if(!module){
        throw new Error('Cannot find module "'+uri+'" @[module: '+callingModule.id+' package: '+callingModule.pkg.name+']');
      }
      module.parent = callingModule;
      return module.call();
    };
  }
  function module(parentId, wrapper){
    var parent = pkgdefs[parentId],
        mod = wrapper(parent),
        cached = false;
    mod.exports = {};
    mod.require = genRequire(mod);
    mod.call = function(){
            if(cached) {
        return mod.exports;
      }
      cached = true;
      global.require = mod.require;
      mod.wrapper(mod, mod.exports, global, nativeBuffer || global.Buffer, global.require);
      return mod.exports;
    };
    if(parent.mainModuleId == mod.id){
      parent.index = mod;
      parent.parents.length === 0 && ( locals.main = mod.call );
    }
    parent.modules.push(mod);
  }
  function pkg(/* [ parentId ...], wrapper */){
    var wrapper = arguments[ arguments.length - 1 ],
        parents = Array.prototype.slice.call(arguments, 0, arguments.length - 1),
        ctx = wrapper(parents);
    if(pkgdefs.hasOwnProperty(ctx.id)){
      throw new Error('Package#'+ctx.id+' "' + ctx.name + '" has duplication of itself.');
    }
    pkgdefs[ctx.id] = ctx;
    pkgmap[ctx.name] = ctx;
    arguments.length == 1 && ( pkgmap.main = ctx );
  }
  function mainRequire(uri){
    return pkgmap.main.index.require(uri);
  }
  function stderr(){
    return lib.process.stderr.content;
  }
  function stdin(){
    return lib.process.stdin.content;
  }
  function stdout(){
    return lib.process.stdout.content;
  }
  return (locals = {
    'lib'        : lib,
    'findPkg'    : findPkg,
    'findModule' : findModule,
    'name'       : 'exampleProject',
    'module'     : module,
    'pkg'        : pkg,
    'packages'   : pkgmap,
    'stderr'     : stderr,
    'stdin'      : stdin,
    'stdout'     : stdout,
    'require'    : mainRequire
});
}(this));
exampleProject.pkg(1, function(parents){
  return {
    'id':9,
    'name':'assert',
    'main':undefined,
    'mainModuleId':'assert',
    'modules':[],
    'parents':parents
  };
});
exampleProject.module(9, function(/* parent */){
  return {
    'id': 'assert',
    'pkg': arguments[0],
    'wrapper': function(module, exports, global, Buffer,require, undefined){
      exports.assert = true;
    }
  };
});
exampleProject.pkg(1, function(parents){
  return {
    'id':3,
    'name':'dependency',
    'main':undefined,
    'mainModuleId':'f',
    'modules':[],
    'parents':parents
  };
});
exampleProject.module(3, function(/* parent */){
  return {
    'id': 'f',
    'pkg': arguments[0],
    'wrapper': function(module, exports, global, Buffer,require, undefined){
      require('subdependency');
exports.parent = module.parent;
exports.f = true;
    }
  };
});
exampleProject.module(3, function(/* parent */){
  return {
    'id': 'g',
    'pkg': arguments[0],
    'wrapper': function(module, exports, global, Buffer,require, undefined){
      exports.g = true;
    }
  };
});
exampleProject.pkg(function(parents){
  return {
    'id':1,
    'name':'example-project',
    'main':undefined,
    'mainModuleId':'a',
    'modules':[],
    'parents':parents
  };
});
exampleProject.module(1, function(/* parent */){
  return {
    'id': 'a',
    'pkg': arguments[0],
    'wrapper': function(module, exports, global, Buffer,require, undefined){
      console.log('Elle creuse encore, cette vieville amie au regard fatigué.');
var mustacheSyntax = "{{ foobar }}",
    assert = require('assert'),
    dependency = require('dependency'),
    sibling = require('sibling'),
    b = require('./b'),
    errorThrown;
errorThrown =  false;
try {
  require('notdeclared');
} catch(err){
  errorThrown = true;
}
if(!errorThrown) throw new Error('"notdeclared" is included');
errorThrown =  false;
try {
  require('./ignored1');
} catch(err){
  errorThrown = true;
}
if(!errorThrown) throw new Error('"ignored1" is included');
errorThrown =  false;
try {
  require('./ignored2');
} catch(err){
  errorThrown = true;
}
if(!errorThrown) throw new Error('"ignored2" is included');
errorThrown =  false;
try {
  require('./ignored3');
} catch(err){
  errorThrown = true;
}
if(!errorThrown) throw new Error('"ignored3" is included');
module.exports = {
  'a':true,
  'dependency': dependency,
  'now': +(new Date),
  'global':global,
  'process':process,
  'Buffer':Buffer,
  'console': console,
  'mustacheSyntax': mustacheSyntax
};
    }
  };
});
exampleProject.module(1, function(/* parent */){
  return {
    'id': 'b',
    'pkg': arguments[0],
    'wrapper': function(module, exports, global, Buffer,require, undefined){
      exports.b = true;
    }
  };
});
exampleProject.module(1, function(/* parent */){
  return {
    'id': 'web',
    'pkg': arguments[0],
    'wrapper': function(module, exports, global, Buffer,require, undefined){
      console.log('this module will be working for only web browsers');
    }
  };
});
exampleProject.pkg(1, 4, function(parents){
  return {
    'id':2,
    'name':'exclude',
    'main':undefined,
    'mainModuleId':'index',
    'modules':[],
    'parents':parents
  };
});
exampleProject.module(2, function(/* parent */){
  return {
    'id': 'index',
    'pkg': arguments[0],
    'wrapper': function(module, exports, global, Buffer,require, undefined){
      exports.exclude = true;
    }
  };
});
exampleProject.pkg(3, function(parents){
  return {
    'id':5,
    'name':'fruits',
    'main':undefined,
    'mainModuleId':'index',
    'modules':[],
    'parents':parents
  };
});
exampleProject.module(5, function(/* parent */){
  return {
    'id': 'index',
    'pkg': arguments[0],
    'wrapper': function(module, exports, global, Buffer,require, undefined){
      module.exports = require('./lib/fruits');
    }
  };
});
exampleProject.module(5, function(/* parent */){
  return {
    'id': 'lib/fruits',
    'pkg': arguments[0],
    'wrapper': function(module, exports, global, Buffer,require, undefined){
      module.exports = ['apple', 'orange'];
    }
  };
});
exampleProject.pkg(3, 1, function(parents){
  return {
    'id':6,
    'name':'sibling',
    'main':undefined,
    'mainModuleId':'n',
    'modules':[],
    'parents':parents
  };
});
exampleProject.module(6, function(/* parent */){
  return {
    'id': 'n',
    'pkg': arguments[0],
    'wrapper': function(module, exports, global, Buffer,require, undefined){
      
exports.n = true;
exports.p = require('./p');
exports.s = require('./s/t');
    }
  };
});
exampleProject.module(6, function(/* parent */){
  return {
    'id': 'p/index',
    'pkg': arguments[0],
    'wrapper': function(module, exports, global, Buffer,require, undefined){
      exports.p = true;
exports.index = true;
    }
  };
});
exampleProject.module(6, function(/* parent */){
  return {
    'id': 'p/r',
    'pkg': arguments[0],
    'wrapper': function(module, exports, global, Buffer,require, undefined){
      require('../s/t');
exports.r = true;
    }
  };
});
exampleProject.module(6, function(/* parent */){
  return {
    'id': 's/t',
    'pkg': arguments[0],
    'wrapper': function(module, exports, global, Buffer,require, undefined){
      exports.t = true;
    }
  };
});
exampleProject.pkg(3, function(parents){
  return {
    'id':4,
    'name':'subdependency',
    'main':undefined,
    'mainModuleId':'i',
    'modules':[],
    'parents':parents
  };
});
exampleProject.module(4, function(/* parent */){
  return {
    'id': 'i',
    'pkg': arguments[0],
    'wrapper': function(module, exports, global, Buffer,require, undefined){
      require('sibling');
exports.i = true;
    }
  };
});
exampleProject.pkg(3, function(parents){
  return {
    'id':7,
    'name':'vegetables',
    'main':undefined,
    'mainModuleId':'lib/index',
    'modules':[],
    'parents':parents
  };
});
exampleProject.module(7, function(/* parent */){
  return {
    'id': 'lib/index',
    'pkg': arguments[0],
    'wrapper': function(module, exports, global, Buffer,require, undefined){
      module.exports = ['tomato', 'potato'];
    }
  };
});
exampleProject.pkg(3, function(parents){
  return {
    'id':8,
    'name':'vehicles',
    'main':undefined,
    'mainModuleId':'index',
    'modules':[],
    'parents':parents
  };
});
exampleProject.module(8, function(/* parent */){
  return {
    'id': 'index',
    'pkg': arguments[0],
    'wrapper': function(module, exports, global, Buffer,require, undefined){
      module.exports = require('./lib/vehicles');
    }
  };
});
exampleProject.module(8, function(/* parent */){
  return {
    'id': 'lib/vehicles',
    'pkg': arguments[0],
    'wrapper': function(module, exports, global, Buffer,require, undefined){
      module.exports = ['car', 'boat', 'truck'];
    }
  };
});
if(typeof module != 'undefined' && module.exports ){
  module.exports = exampleProject;
  if( !module.parent ){
    exampleProject.main();
  }
}
